File: src/test/java/dev/nytweetdeck/xapi/graphql/AuthenticatedGraphQlClientTest.java
package dev.nytweetdeck.xapi.graphql;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import dev.nytweetdeck.account.AccountSecrets;
import dev.nytweetdeck.account.AccountStore;
import dev.nytweetdeck.xapi.http.XApiHttpException;
import dev.nytweetdeck.xapi.http.XClientTransactionIdService;
import dev.nytweetdeck.xapi.profile.XApiProfile;
import dev.nytweetdeck.xapi.profile.XApiProfile.OperationType;
import dev.nytweetdeck.xapi.profile.XApiProfileService;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URLDecoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Flow;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import tools.jackson.databind.json.JsonMapper;

class AuthenticatedGraphQlClientTest {

    @TempDir
    Path temporaryDirectory;

    private HttpServer server;
    private URI baseUri;

    @BeforeEach
    void startServer() throws IOException {
        server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        server.start();
        baseUri = URI.create("http://127.0.0.1:" + server.getAddress().getPort());
    }

    @AfterEach
    void stopServer() {
        server.stop(0);
    }

    @Test
    void sendsVersionedQueryWithSelectedFeaturesAndWebSession() throws Exception {
        var rawQuery = new AtomicReference<String>();
        var authorization = new AtomicReference<String>();
        var cookie = new AtomicReference<String>();
        var csrfToken = new AtomicReference<String>();
        server.createContext("/graphql/op-id/HomeTimeline", exchange -> {
            rawQuery.set(exchange.getRequestURI().getRawQuery());
            authorization.set(exchange.getRequestHeaders().getFirst("Authorization"));
            cookie.set(exchange.getRequestHeaders().getFirst("Cookie"));
            csrfToken.set(exchange.getRequestHeaders().getFirst("X-CSRF-Token"));
            respond(exchange, 200, "{\"data\":{\"timeline\":{}}}");
        });
        var client = createClient();

        var result = client.execute("account-1", "homeForYou", Map.of("count", 20));

        var query = parseQuery(rawQuery.get());
        assertThat(query.get("variables")).contains("\"count\":20");
        assertThat(query.get("features")).isEqualTo("{\"feature_enabled\":true}");
        assertThat(query.get("fieldToggles"))
                .contains("\"withArticlePlainText\":false", "\"withPayments\":false");
        assertThat(authorization.get()).isEqualTo("Bearer web-bearer");
        assertThat(cookie.get()).contains("auth_token=web-auth", "ct0=web-csrf");
        assertThat(csrfToken.get()).isEqualTo("web-csrf");
        assertThat(result.rawJson()).contains("timeline");
        assertThat(result.toString()).doesNotContain("timeline");
    }

    @Test
    void rejectsGraphQlErrorWithoutReturningRawDetails() throws Exception {
        server.createContext("/graphql/op-id/HomeTimeline", exchange ->
                respond(exchange, 200, "{\"errors\":[{\"message\":\"private detail\"}]}"));
        var client = createClient();

        assertThatThrownBy(() -> client.execute("account-1", "homeForYou", Map.of()))
                .isInstanceOf(XApiHttpException.class)
                .hasMessageContaining("エラーを返しました")
                .hasMessageNotContaining("private detail");
    }

    @Test
    void acceptsPartialGraphQlDataWhenNonFatalErrorsAreAlsoPresent() throws Exception {
        server.createContext("/graphql/op-id/HomeTimeline", exchange -> respond(
                exchange,
                200,
                "{\"data\":{\"timeline\":{}},\"errors\":[{\"message\":\"one item unavailable\"}]}"));
        var client = createClient();

        var result = client.execute("account-1", "homeForYou", Map.of());

        assertThat(result.rawJson()).contains("timeline");
    }

    @Test
    void sendsSearchTimelineAsPostWithJsonPayload() throws Exception {
        var client = createClient("search", "SearchTimeline");

        var request = client
                .prepareRequest(
                        "account-1",
                        "search",
                        Map.of("rawQuery", "NyTweetDeck", "count", 20))
                .request();

        var payload = JsonMapper.builder().build().readTree(readBody(request));
        assertThat(request.method()).isEqualTo("POST");
        assertThat(request.uri().getRawQuery()).isNull();
        assertThat(request.headers().firstValue("Content-Type")).contains("application/json");
        assertThat(payload.get("variables").get("rawQuery").asString()).isEqualTo("NyTweetDeck");
        assertThat(payload.get("features").get("feature_enabled").asBoolean()).isTrue();
        assertThat(payload.get("fieldToggles").get("withArticlePlainText").asBoolean()).isFalse();
    }

    @Test
    void sendsTheNyTweetDeckUiLanguageToXWeb() throws Exception {
        var client = createClient();

        var request = client.prepareRequest(
                        "account-1", "homeForYou", Map.of("count", 20), "fr-CA")
                .request();

        assertThat(request.headers().firstValue("X-Twitter-Client-Language"))
                .contains("fr-ca");
        assertThatThrownBy(() -> client.prepareRequest(
                        "account-1", "homeForYou", Map.of(), "ja\r\nInjected: yes"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("表示言語");
    }

    @Test
    void includesTheResolvedQueryIdInMutationPayloads() throws Exception {
        var client = createClient("bookmark", "CreateBookmark", OperationType.MUTATION);

        var request = client
                .prepareRequest("account-1", "bookmark", Map.of("tweet_id", "123"))
                .request();
        var payload = JsonMapper.builder().build().readTree(readBody(request));

        assertThat(request.method()).isEqualTo("POST");
        assertThat(request.headers().firstValue("X-Client-Transaction-Id"))
                .contains("test-transaction-id");
        assertThat(payload.get("queryId").asString()).isEqualTo("op-id");
        assertThat(payload.get("variables").get("tweet_id").asString()).isEqualTo("123");
        assertThat(payload.get("features")).isNull();
    }

    @Test
    void refreshesTheWebTransactionMaterialOnceWhenXMisreportsMutationCode344()
            throws Exception {
        var requests = new AtomicInteger();
        server.createContext("/graphql/op-id/CreateRetweet", exchange -> {
            if (requests.incrementAndGet() == 1) {
                respond(exchange, 200, "{\"errors\":[{\"code\":344}]}");
            } else {
                respond(exchange, 200, "{\"data\":{\"create_retweet\":{}}}");
            }
        });
        var transactions = new RecordingTransactionService();
        var client = createClient(
                "repost", "CreateRetweet", OperationType.MUTATION, transactions);

        var result = client.execute("account-1", "repost", Map.of("tweet_id", "123"));

        assertThat(result.rawJson()).contains("create_retweet");
        assertThat(requests).hasValue(2);
        assertThat(transactions.invalidations).hasValue(1);
        assertThat(transactions.generations).hasValue(2);
    }

    private AuthenticatedGraphQlClient createClient() throws Exception {
        return createClient("homeForYou", "HomeTimeline");
    }

    private AuthenticatedGraphQlClient createClient(String purpose, String operationName)
            throws Exception {
        return createClient(purpose, operationName, OperationType.QUERY);
    }

    private AuthenticatedGraphQlClient createClient(
            String purpose, String operationName, OperationType operationType)
            throws Exception {
        return createClient(
                purpose,
                operationName,
                operationType,
                new XClientTransactionIdService(null) {
                    @Override
                    public String generate(String method, URI requestUri) {
                        return "test-transaction-id";
                    }
                });
    }

    private AuthenticatedGraphQlClient createClient(
            String purpose,
            String operationName,
            OperationType operationType,
            XClientTransactionIdService transactionIdService)
            throws Exception {
        var jsonMapper = JsonMapper.builder().build();
        var operation = new XApiProfile.GraphQlOperation(
                purpose, "op-id", operationName, operationType);
        var profile = new XApiProfile(
                "x-web",
                "current",
                0,
                baseUri,
                baseUri.resolve("/graphql"),
                Map.of("X-Twitter-Client", "TwitterWebClient"),
                Map.of(),
                java.util.List.of("feature_enabled"),
                Map.of(purpose, operation));
        var profileService = new XApiProfileService(jsonMapper) {
            @Override
            public XApiProfile profile() {
                return profile;
            }

            @Override
            public XApiProfile.GraphQlOperation requireOperation(String purpose) {
                return profile.graphqlOperations().get(purpose);
            }

            @Override
            public Map<String, Boolean> selectFeatures(XApiProfile.GraphQlOperation ignored) {
                return Map.of("feature_enabled", true);
            }
        };
        var accountStore = new AccountStore(
                jsonMapper, temporaryDirectory.resolve("accounts.json"));
        accountStore.addOrReplace(AccountSecrets.webSession(
                "account-1", "42", "alice", "Alice", "web-bearer", "web-auth", "web-csrf"));
        return new AuthenticatedGraphQlClient(
                HttpClient.newHttpClient(),
                jsonMapper,
                profileService,
                accountStore,
                transactionIdService);
    }

    private static Map<String, String> parseQuery(String rawQuery) {
        var pairs = new java.util.LinkedHashMap<String, String>();
        for (String part : rawQuery.split("&")) {
            var separator = part.indexOf('=');
            pairs.put(
                    URLDecoder.decode(part.substring(0, separator), StandardCharsets.UTF_8),
                    URLDecoder.decode(part.substring(separator + 1), StandardCharsets.UTF_8));
        }
        return pairs;
    }

    private static String readBody(HttpRequest request) {
        var publisher = request.bodyPublisher().orElseThrow();
        var subscriber = new BodySubscriber();
        publisher.subscribe(subscriber);
        return subscriber.body().join();
    }

    private static final class BodySubscriber implements Flow.Subscriber<ByteBuffer> {
        private final ByteArrayOutputStream output = new ByteArrayOutputStream();
        private final CompletableFuture<String> body = new CompletableFuture<>();

        @Override
        public void onSubscribe(Flow.Subscription subscription) {
            subscription.request(Long.MAX_VALUE);
        }

        @Override
        public void onNext(ByteBuffer item) {
            var bytes = new byte[item.remaining()];
            item.get(bytes);
            output.writeBytes(bytes);
        }

        @Override
        public void onError(Throwable throwable) {
            body.completeExceptionally(throwable);
        }

        @Override
        public void onComplete() {
            body.complete(output.toString(StandardCharsets.UTF_8));
        }

        CompletableFuture<String> body() {
            return body;
        }
    }

    private static final class RecordingTransactionService
            extends XClientTransactionIdService {
        private final AtomicInteger generations = new AtomicInteger();
        private final AtomicInteger invalidations = new AtomicInteger();

        private RecordingTransactionService() {
            super(null);
        }

        @Override
        public String generate(String method, URI requestUri) {
            return "test-transaction-id-" + generations.incrementAndGet();
        }

        @Override
        public synchronized void invalidate() {
            invalidations.incrementAndGet();
        }
    }

    private static void respond(HttpExchange exchange, int status, String body) throws IOException {
        var bytes = body.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.getResponseHeaders().set("Connection", "close");
        exchange.sendResponseHeaders(status, bytes.length);
        try (var output = exchange.getResponseBody()) {
            output.write(bytes);
        }
    }
}

File: android/app/src/test/java/dev/nytweetdeck/android/xapi/AuthenticatedGraphQlClientTest.kt
package dev.nytweetdeck.android.xapi

import okhttp3.HttpUrl
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Protocol
import okhttp3.Request
import okhttp3.Response
import okhttp3.ResponseBody.Companion.toResponseBody
import okio.Buffer
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

class AuthenticatedGraphQlClientTest {
    @Test
    fun officialQueryCarriesWebSessionAndTransactionHeaders() {
        val interceptor = SequenceInterceptor(ResponseSpec(200, """{"data":{"ok":true}}"""))
        val transactions = RecordingTransactionIdService()
        val client = client(profile(), interceptor, transactions)

        client.execute(
            credentials = credentials(),
            purpose = "homeForYou",
            variables = mapOf("count" to 20, "includePromotedContent" to false),
            language = "ja-JP",
        )

        val request = interceptor.singleRequest()
        assertEquals("Bearer bearer-fixture", request.header("Authorization"))
        assertEquals("csrf-fixture", request.header("X-CSRF-Token"))
        assertEquals("ja-jp", request.header("X-Twitter-Client-Language"))
        assertEquals("transaction-1", request.header(TRANSACTION_HEADER))
        assertNotNull(request.url.queryParameter("variables"))
        assertNotNull(request.url.queryParameter("features"))
        assertEquals(listOf("GET /i/api/graphql/operation-id/HomeTimeline"), transactions.generated)
    }

    @Test
    fun mutationCarriesTheHeaderAndRefreshesExactlyOnceForCode344() {
        val interceptor = SequenceInterceptor(
            ResponseSpec(200, """{"errors":[{"code":344}]}"""),
            ResponseSpec(200, """{"data":{"bookmark":true}}"""),
        )
        val transactions = RecordingTransactionIdService()
        val client = client(
            profile(
                operations = mapOf(
                    "bookmark" to operation(
                        id = "mutation-id",
                        name = "CreateBookmark",
                        type = XApiProfile.OperationType.MUTATION,
                    ),
                ),
            ),
            interceptor,
            transactions,
        )

        val result = client.execute(credentials(), "bookmark", mapOf("tweet_id" to "123"), "ja")

        assertTrue(result.contains("bookmark"))
        assertEquals(2, interceptor.requests.size)
        assertEquals("transaction-1", interceptor.requests[0].header(TRANSACTION_HEADER))
        assertEquals("transaction-2", interceptor.requests[1].header(TRANSACTION_HEADER))
        assertEquals(1, transactions.invalidations)
        assertEquals(
            listOf(
                "POST /i/api/graphql/mutation-id/CreateBookmark",
                "POST /i/api/graphql/mutation-id/CreateBookmark",
            ),
            transactions.generated,
        )
        assertTrue(interceptor.requests[0].bodyText().contains("\"queryId\":\"mutation-id\""))
        assertFalse(interceptor.requests[0].bodyText().contains("fieldToggles"))
    }

    @Test
    fun postQueryCarriesTheHeaderUsingItsActualMethodAndPath() {
        val interceptor = SequenceInterceptor(ResponseSpec(200, """{"data":{"search":{}}}"""))
        val transactions = RecordingTransactionIdService()
        val client = client(
            profile(
                operations = mapOf(
                    "search" to operation("search-id", "SearchTimeline"),
                ),
            ),
            interceptor,
            transactions,
        )

        client.execute(credentials(), "search", mapOf("rawQuery" to "NyTD"), "ja")

        val request = interceptor.singleRequest()
        assertEquals("POST", request.method)
        assertEquals("transaction-1", request.header(TRANSACTION_HEADER))
        assertEquals(listOf("POST /i/api/graphql/search-id/SearchTimeline"), transactions.generated)
    }

    @Test
    fun transactionGenerationFailureIsRetriedWithoutSendingAnApiRequest() {
        val interceptor = SequenceInterceptor(ResponseSpec(200, """{"data":{"ok":true}}"""))
        val transactions = RecordingTransactionIdService(failuresBeforeSuccess = 2)
        val client = client(profile(), interceptor, transactions)

        val error = assertThrows(XApiException::class.java) {
            client.execute(credentials(), "homeForYou", mapOf("count" to 20), "ja")
        }

        assertEquals(2, transactions.generated.size)
        assertEquals(1, transactions.invalidations)
        assertTrue(interceptor.requests.isEmpty())
        assertTrue(requireNotNull(error.message).contains("Web署名を生成できません"))
        assertFalse(requireNotNull(error.message).contains("fixture transaction failure"))
    }

    @Test
    fun oneTransactionGenerationFailureRecoversBeforeTheOnlyApiRequest() {
        val interceptor = SequenceInterceptor(ResponseSpec(200, """{"data":{"ok":true}}"""))
        val transactions = RecordingTransactionIdService(failuresBeforeSuccess = 1)
        val client = client(profile(), interceptor, transactions)

        client.execute(credentials(), "homeForYou", mapOf("count" to 20), "ja")

        assertEquals(2, transactions.generated.size)
        assertEquals(1, transactions.invalidations)
        assertEquals(1, interceptor.requests.size)
        assertEquals("transaction-2", interceptor.singleRequest().header(TRANSACTION_HEADER))
    }

    @Test
    fun customNonOfficialProfilesKeepTheExistingUnsignedRequestBehavior() {
        val interceptor = SequenceInterceptor(ResponseSpec(200, """{"data":{"ok":true}}"""))
        val transactions = RecordingTransactionIdService()
        val client = client(
            profile(graphqlBaseUrl = "https://example.invalid/graphql"),
            interceptor,
            transactions,
        )

        client.execute(credentials(), "homeForYou", mapOf("count" to 20), "ja")

        val request = interceptor.singleRequest()
        assertNull(request.header(TRANSACTION_HEADER))
        assertTrue(transactions.generated.isEmpty())
    }

    @Test
    fun responseWithOnlyGraphQlErrorsIsRejected() {
        val interceptor = SequenceInterceptor(ResponseSpec(200, """{"errors":[{"message":"failed"}]}"""))
        val client = client(profile(), interceptor, RecordingTransactionIdService())

        assertThrows(XApiException::class.java) {
            client.execute(credentials(), "homeForYou", mapOf("count" to 20), "ja")
        }
    }

    private fun client(
        profile: XApiProfile,
        interceptor: SequenceInterceptor,
        transactions: RecordingTransactionIdService,
    ): AuthenticatedGraphQlClient = AuthenticatedGraphQlClient(
        client = OkHttpClient.Builder().addInterceptor(interceptor).build(),
        profile = profile,
        userAgent = "NyTD-Test-UA",
        transactionIdService = transactions,
    )

    private fun credentials() = XSessionCredentials("bearer-fixture", "auth-fixture", "csrf-fixture")

    private fun profile(
        graphqlBaseUrl: String = "https://x.com/i/api/graphql",
        operations: Map<String, XApiProfile.GraphQlOperation> = mapOf(
            "homeForYou" to operation("operation-id", "HomeTimeline"),
        ),
    ) = XApiProfile(
        graphqlBaseUrl = graphqlBaseUrl,
        featureKeys = listOf("feature-a"),
        featureDefaults = mapOf("feature-a" to true),
        operations = operations,
    )

    private fun operation(
        id: String,
        name: String,
        type: XApiProfile.OperationType = XApiProfile.OperationType.QUERY,
    ) = XApiProfile.GraphQlOperation(
        operationId = id,
        operationName = name,
        type = type,
        featureKeys = emptyList(),
        fieldToggles = emptyList(),
    )

    private class RecordingTransactionIdService(
        private var failuresBeforeSuccess: Int = 0,
    ) : XClientTransactionIdService(
        XClientTransactionIdService.AssetFetcher { _, _ -> "" },
    ) {
        val generated = mutableListOf<String>()
        var invalidations = 0

        override fun generate(method: String, requestUrl: HttpUrl): String {
            generated += "$method ${requestUrl.encodedPath}"
            if (failuresBeforeSuccess > 0) {
                failuresBeforeSuccess -= 1
                throw XApiException("fixture transaction failure", 502)
            }
            return "transaction-${generated.size}"
        }

        override fun invalidate() {
            invalidations += 1
        }
    }

    private class SequenceInterceptor(
        vararg responses: ResponseSpec,
    ) : Interceptor {
        private val responses = responses.toList()
        private var nextResponse = 0
        val requests = mutableListOf<Request>()

        override fun intercept(chain: Interceptor.Chain): Response {
            val request = chain.request()
            requests += request
            val response = responses[nextResponse.coerceAtMost(responses.lastIndex)]
            nextResponse += 1
            return Response.Builder()
                .request(request)
                .protocol(Protocol.HTTP_1_1)
                .code(response.statusCode)
                .message(if (response.statusCode in 200..299) "OK" else "Rejected")
                .body(response.body.toResponseBody("application/json".toMediaType()))
                .build()
        }

        fun singleRequest(): Request = require(requests.size == 1) { "expected one request" }.let { requests.single() }
    }

    private data class ResponseSpec(
        val statusCode: Int,
        val body: String,
    )

    private fun Request.bodyText(): String {
        val buffer = Buffer()
        requireNotNull(body).writeTo(buffer)
        return buffer.readUtf8()
    }

    private companion object {
        const val TRANSACTION_HEADER = "X-Client-Transaction-Id"
    }
}
